src/aspire/abinitio/commonline_lud.py

```diff
@@ -1,32 +1,551 @@
+
+                nD = dH > self.EPS
+                dH = dH[nD]
+                nev = np.count_nonzero(nD)
+                W = V[:, nD] @ np.diag(dH) @ V[:, nD].T + H if nD.any() else H
+
+            ############
+            # Update G #
+            ############
+            G = (1 - self.gam) * G + self.gam * self.mu * (W - H)
+
+            # Check optimality
+            resi = self._compute_AX(G) - b
+            pinf = np.linalg.norm(resi) / max(np.linalg.norm(b), 1)
+
+            dinf_term = S + W + ATy
+            if self.alpha is not None:
+                dinf_term -= Z
+            dinf = np.linalg.norm(dinf_term, "fro") / max(np.linalg.norm(S, np.inf), 1)
+
+            logger.info(
+                f"Iteration: {itr}, residual: {max(pinf, dinf)}, target: {self.tol}"
+            )
+            if max(pinf, dinf) <= self.tol:
+                return G
+
+            # Update mu adaptively
+            if self.adp_mu:
+                if pinf / dinf <= self.delta_mu_l:
+                    itmu_pinf = itmu_pinf + 1
+                    itmu_dinf = 0
+                    if itmu_pinf > self.max_mu_itr:
+                        self.mu = max(self.mu * self.inc_mu, self.mu_min)
+                        itmu_pinf = 0
+                elif pinf / dinf > self.delta_mu_u:
+                    itmu_dinf = itmu_dinf + 1
+                    itmu_pinf = 0
+                    if itmu_dinf > self.max_mu_itr:
+                        self.mu = min(self.mu * self.dec_mu, self.mu_max)
+                        itmu_dinf = 0
+
+        return G
+
+    def _compute_Z(self, S, W, ATy, G, zz, itr, kk, nev):
+        """
+        Update ADMM subproblem for enforcing the spectral norm constraint.
+
+        :param S: A 2*n_img x 2*n_img symmetric matrix representing the fidelity term.
+        :param W: A 2*n_img x 2*n_img array, primary ADMM subproblem matrix.
+        :param ATy: A 2*n_img x 2*n_img array.
+        :param G: Current value of the 2*n_img x 2*n_img optimization solution matrix.
+        :param zz: eigenvalues from previous iteration.
+        :param itr: ADMM loop iteration.
+        :param kk: Number of eigenvalues of Z to use to enforce spectral norm constraint.
+        :param nev: Number of eigenvalues of W used in previous iteration of ADMM.
+
+        :returns:
+            - Z, Updated 2*n_img x 2*n_img matrix for spectral norm constraint ADMM subproblem.
+            - kk, Number of eigenvalues of Z to use to enforce spectral norm constraint in next iteration.
+            - nev, Number of eigenvalues of W to use in this iteration of ADMM.
+        """
+        lambda_ = self.alpha * self.n_img  # Spectral norm bound
+        B = S + W + ATy + G / self.mu
+        B = (B + B.T) / 2
+
+        if not self.adp_proj:
+            U, pi = np.linalg.eigh(B)
+        else:
+            # Determine number of eigenvalues to compute for adaptive projection
+            if itr == 0:
```

Comment:
Why not just supply `kk = None` in the first iteration and check for that. A bit awkward to pass the iteration number to this inner function.

Reply:
Good point. That's much cleaner. Done. Thanks!